{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests \n",
    "from bs4 import BeautifulSoup\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://link.springer.com/'\n",
    "data = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Simple and multiple linear regression models are considered between variables whose “values” are convex compact random sets in \\\\({\\\\mathbb{R}^p}\\\\) , (that is, hypercubes, spheres, and so on). We analyze such models within a set-arithmetic approach. Contrary to what happens for random variables, the least squares optimal solutions for the basic affine transformation model do not produce suitable estimates for the linear regression model. First, we derive least squares estimators for the simple linear regression model and examine them from a theoretical perspective. Moreover, the multiple linear regression model is dealt with and a stepwise algorithm is developed in order to find the estimates in this case. The particular problem of the linear regression with interval-valued data is also considered and illustrated by means of a real-life example.'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper.find('div', attrs={'class':'c-article-section__content', 'id':'Abs1-content'}).text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n"
     ]
    }
   ],
   "source": [
    "r = requests.get(url+'journal/11634/volumes-and-issues')\n",
    "soup = BeautifulSoup(r.text, 'html.parser')\n",
    "results = soup('li', class_='c-list-group__item')\n",
    "i=0\n",
    "for result in results:\n",
    "    year = re.findall('\\d\\d\\d\\d', str(result.find('a')))[0]\n",
    "    issue_link = result.find('a')['href']\n",
    "#     issue_link = result.a['href']\n",
    "    r = requests.get(url+issue_link)\n",
    "    soup = BeautifulSoup(r.text, 'html.parser')\n",
    "    publications = soup.find_all('div', attrs={'class':'toc-item no-access', 'class':'toc-item'})\n",
    "    \n",
    "    for publication in publications:\n",
    "#         if publication.find('p', 'content-type').text == 'OriginalPaper':\n",
    "#         if publication.find('p', 'content-type').text in ['OriginalPaper', 'Authors Of Articles', 'Article']:\n",
    "\n",
    "        paper_link = publication.find('a')['href']\n",
    "        r = requests.get(url+paper_link)\n",
    "        paper = BeautifulSoup(r.text, 'html.parser')\n",
    "        description=paper.find('h2', attrs={'class':\n",
    "                        'c-article-section__title js-section-title js-c-reading-companion-sections-item'})\n",
    "        if description:\n",
    "            if 'Abstract' in description.text:\n",
    "\n",
    "                abstract = paper.find('div', attrs={'class':'c-article-section__content', 'id':'Abs1-content'})\n",
    "                keywords = paper.find('ul',class_=\"c-article-subject-list\")\n",
    "\n",
    "                if abstract:\n",
    "                    if abstract.text:\n",
    "                        i+=1\n",
    "                        print(i)\n",
    "\n",
    "                        abstract = re.sub(r'\\n', '', abstract.text)\n",
    "\n",
    "                        headline = paper.find('h1',class_='c-article-title').text\n",
    "                        headline = re.sub('\\n', '', headline)\n",
    "                        if keywords:\n",
    "                            keywords = keywords.find_all('span', itemprop=\"about\")\n",
    "                            keywords = \", \".join([keyword.text for keyword in keywords])\n",
    "                            keywords = re.sub(r'\\n', '', keywords)\n",
    "                        else:\n",
    "                            keywords=''\n",
    "                        data.append((headline, abstract, keywords, url+paper_link, year))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper name</th>\n",
       "      <th>abstract</th>\n",
       "      <th>keywords</th>\n",
       "      <th>link</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seemingly unrelated clusterwise linear regression</td>\n",
       "      <td>Linear regression models based on finite Gauss...</td>\n",
       "      <td>EM algorithm, Envelope function, Identifiabili...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Semiparametric mixtures of regressions with si...</td>\n",
       "      <td>In this article, we propose two classes of sem...</td>\n",
       "      <td>EM algorithm, Kernel regression, Mixture regre...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gaussian parsimonious clustering models with c...</td>\n",
       "      <td>We consider model-based clustering methods for...</td>\n",
       "      <td>Model-based clustering, Mixtures of experts, E...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A robust approach to model-based classificatio...</td>\n",
       "      <td>In a standard classification framework a set o...</td>\n",
       "      <td>Model-based classification, Label noise, Outli...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mixture modeling of data with multiple partial...</td>\n",
       "      <td>In this paper, a new flexible approach to mode...</td>\n",
       "      <td>Finite mixture models, EM algorithm, Right-cen...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>Adaptive dissimilarity index for measuring tim...</td>\n",
       "      <td>The most widely used measures of time series p...</td>\n",
       "      <td>Classification, Time Series, Fréchet distance,...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>Generalized mixture models, semi-supervised le...</td>\n",
       "      <td>In this paper, we discuss generalized mixture ...</td>\n",
       "      <td>Mixture Model, Label Data, Unlabeled Sample, A...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>Mixture-model-based signal denoising</td>\n",
       "      <td>This paper proposes a new signal denoising met...</td>\n",
       "      <td>Denoising, Asymmetrical noise, Regression, Gau...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>Galois closed entity sets and k-balls of quasi...</td>\n",
       "      <td>Quasi-ultrametric multi-way dissimilarities an...</td>\n",
       "      <td>Closure operator, Cluster analysis, Galois con...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>Least squares estimation of linear regression ...</td>\n",
       "      <td>Simple and multiple linear regression models a...</td>\n",
       "      <td>Linear regression model, Convex compact random...</td>\n",
       "      <td>https://link.springer.com//article/10.1007/s11...</td>\n",
       "      <td>1163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>332 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            paper name  \\\n",
       "0    Seemingly unrelated clusterwise linear regression   \n",
       "1    Semiparametric mixtures of regressions with si...   \n",
       "2    Gaussian parsimonious clustering models with c...   \n",
       "3    A robust approach to model-based classificatio...   \n",
       "4    Mixture modeling of data with multiple partial...   \n",
       "..                                                 ...   \n",
       "327  Adaptive dissimilarity index for measuring tim...   \n",
       "328  Generalized mixture models, semi-supervised le...   \n",
       "329               Mixture-model-based signal denoising   \n",
       "330  Galois closed entity sets and k-balls of quasi...   \n",
       "331  Least squares estimation of linear regression ...   \n",
       "\n",
       "                                              abstract  \\\n",
       "0    Linear regression models based on finite Gauss...   \n",
       "1    In this article, we propose two classes of sem...   \n",
       "2    We consider model-based clustering methods for...   \n",
       "3    In a standard classification framework a set o...   \n",
       "4    In this paper, a new flexible approach to mode...   \n",
       "..                                                 ...   \n",
       "327  The most widely used measures of time series p...   \n",
       "328  In this paper, we discuss generalized mixture ...   \n",
       "329  This paper proposes a new signal denoising met...   \n",
       "330  Quasi-ultrametric multi-way dissimilarities an...   \n",
       "331  Simple and multiple linear regression models a...   \n",
       "\n",
       "                                              keywords  \\\n",
       "0    EM algorithm, Envelope function, Identifiabili...   \n",
       "1    EM algorithm, Kernel regression, Mixture regre...   \n",
       "2    Model-based clustering, Mixtures of experts, E...   \n",
       "3    Model-based classification, Label noise, Outli...   \n",
       "4    Finite mixture models, EM algorithm, Right-cen...   \n",
       "..                                                 ...   \n",
       "327  Classification, Time Series, Fréchet distance,...   \n",
       "328  Mixture Model, Label Data, Unlabeled Sample, A...   \n",
       "329  Denoising, Asymmetrical noise, Regression, Gau...   \n",
       "330  Closure operator, Cluster analysis, Galois con...   \n",
       "331  Linear regression model, Convex compact random...   \n",
       "\n",
       "                                                  link  year  \n",
       "0    https://link.springer.com//article/10.1007/s11...  1163  \n",
       "1    https://link.springer.com//article/10.1007/s11...  1163  \n",
       "2    https://link.springer.com//article/10.1007/s11...  1163  \n",
       "3    https://link.springer.com//article/10.1007/s11...  1163  \n",
       "4    https://link.springer.com//article/10.1007/s11...  1163  \n",
       "..                                                 ...   ...  \n",
       "327  https://link.springer.com//article/10.1007/s11...  1163  \n",
       "328  https://link.springer.com//article/10.1007/s11...  1163  \n",
       "329  https://link.springer.com//article/10.1007/s11...  1163  \n",
       "330  https://link.springer.com//article/10.1007/s11...  1163  \n",
       "331  https://link.springer.com//article/10.1007/s11...  1163  \n",
       "\n",
       "[332 rows x 5 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(data, columns=['paper name', 'abstract', 'keywords', 'link', 'year'])\n",
    "df.to_csv('papers.csv', index=False, encoding='utf-8')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "topics_dict = {'1.1.1. k-means': None,\n",
    " '1.1.2. number of clusters': None,\n",
    " '1.1.3. self organised map': ['som', 'kohonen map'],\n",
    " '1.1.4. affinity propagation': None,\n",
    " '1.1.5. block-modeling': None,\n",
    " '1.1.6. semi-supervised clustering': ['semi-supervised learning', 'labeled data'],\n",
    " '1.2. one cluster': ['single cluster'],\n",
    " '1.3.1. agglomerative': None,\n",
    " '1.3.2. divisive': None,\n",
    " '1.3.3. decision tree': ['classification tree', 'conceptual clustering'],\n",
    " '1.4. fuzzy clustering': ['soft clustering', 'fuzzy partition'],\n",
    " '1.5.1. density function': ['distribution'],\n",
    " '1.5.2. mixture': None,\n",
    " '1.5.3. expectation maximization': ['maximum likelihood'],\n",
    " '1.5.4. markov models': ['stochastic model', 'hidden markov model', 'markov decision process'],\n",
    " '1.6. cluster-wise regression': ['regression clustering'],\n",
    " '1.7. nature-inspired clustering': ['evolutionary algorithm', 'genetic algorithm', 'particle swarm', 'simulated annealing'],\n",
    " '1.8. consensus clustering': ['ensemble clustering'],\n",
    " '1.9. other clustering': None,\n",
    " '2.1.1. additive clustering': None,\n",
    " '2.1.2. graph mining': ['pattern mining'],\n",
    " '2.1.3. community detection': ['clique'],\n",
    " '2.1.4. spectral clustering': ['normalized cut', 'laplacian normalization'],\n",
    " '2.1.5. maximum spanning tree': ['minimum spanning tree'],\n",
    " '2.1.6. single linkage': ['nearest neighbor'],\n",
    " '2.1.7. complete linkage': ['farthest neighbor'],\n",
    " '2.1.8. kernel function': None,\n",
    " '2.2. distance': ['dissimilarity', 'metric'],\n",
    " '2.3. tree': ['hierarchy', 'dendrogram'],\n",
    " '2.4. preference': ['ordering', 'judgement'],\n",
    " '2.5. incomplete': ['missing', 'gap'],\n",
    " '2.6. spatial data': ['geospatial data'],\n",
    " '2.7. temporal data': ['dynamic', 'time series'],\n",
    " '2.8. three-way': ['tensor', 'tricluster'],\n",
    " '2.9.1. document filtering': ['document summarization'],\n",
    " '2.9.2. topic modeling': ['latent allocation'],\n",
    " '2.9.3. information retrieval': ['document retrieval', 'search', 'tf-idf coding'],\n",
    " '2.9.4. text mining': ['information extraction'],\n",
    " '2.10.1. segmentation': None,\n",
    " '2.10.2. skeleton': ['framework', 'blueprint'],\n",
    " '2.10.3. edge detection': None,\n",
    " '2.11.1. nominal features': ['categorical', 'character', 'attribute'],\n",
    " '2.11.2. binary features': ['dummy variable'],\n",
    " '2.11.3. data coding': ['data encoding', 'quantization'],\n",
    " '2.11.4. feature weighting': ['feature selection', 'variable weighting', 'variable selection'],\n",
    " '2.11.5. data normalization': ['standardization', 'preprocessing'],\n",
    " '3.1. ranking': ['ordering', 'seriation'],\n",
    " '3.2. scaling': ['normalization', 'multidimensional scaling'],\n",
    " '3.3.1. discriminant analysis': None,\n",
    " '3.3.2. support vector machine': ['svm'],\n",
    " '3.3.3. nearest neighbor': None,\n",
    " '3.3.4. neural network': ['feedforward network', 'backpropagation', 'deep learning', 'learning rate'],\n",
    " '3.3.5. regularizer': ['generalization'],\n",
    " '3.3.6. ensemble': ['bagging', 'boosting', 'random forest'],\n",
    " '3.4.1. singular value decomposition': ['svd'],\n",
    " '3.4.2. principal component analysis': None,\n",
    " '3.4.3. correspondence analysis': ['dual scaling'],\n",
    " '3.4.4. independent component': None,\n",
    " '3.6. biclustering': ['coclustering', 'two-mode clustering'],\n",
    " '3.7.1. biplot': None,\n",
    " '3.7.2. dendrogram': None,\n",
    " '3.7.3. cloud': None,\n",
    " '3.7.4. parallel coordinates': None,\n",
    " '3.8.1. non-negative decomposition': None,\n",
    " '3.8.2. archetype analysis': None,\n",
    " '3.9. anomaly detection': ['outlier'],\n",
    " '3.10.1. contingency table': None,\n",
    " '3.10.2. comparing partitions': ['rand index'],\n",
    " '3.10.3. association index': ['chi-squared'],\n",
    " '3.11.1. linear regression': ['correlation coefficient'],\n",
    " '3.11.2. non-linear regression': ['power law'],\n",
    " '3.11.3. spline': None,\n",
    " '3.12. formal concept': ['galois association'],\n",
    " '4.1.1. mixture of distributions': None,\n",
    " '4.1.2. bayesian': None,\n",
    " '4.1.3. maximum likelihood': None,\n",
    " '4.1.4. markov-chain monte carlo methods': None,\n",
    " '4.2. geometry': None,\n",
    " '4.3. optimisation': ['minimization, maximization'],\n",
    " '4.4. cluster structures': None,\n",
    " '4.5.1. cross-validation': None,\n",
    " '4.5.2. resampling': ['random sample'],\n",
    " '4.5.3. leave-one-out': ['jackknife'],\n",
    " '4.5.4. bootstrap': None,\n",
    " '4.5.5. data generation': ['synthetic data'],\n",
    " '4.6.1. accuracy': ['precision', 'sensitivity'],\n",
    " '4.6.2. false positive': ['false negative', 'true positive', 'false positive', 'f-score'],\n",
    " '4.6.3. cluster recovery': ['ground truth', 'kappa statistic'],\n",
    " '4.6.4. receiver operating characteristic': ['area under the curve', 'roc curve, auc'],\n",
    " '4.6.5. quadratic error': ['mean squared error', 'squared error', 'inertia'],\n",
    " '5.1. classification': ['typology', 'taxonomy'],\n",
    " '5.2. evolution': ['phylogeny', 'darwin'],\n",
    " '5.3. psychology': None,\n",
    " '5.4. antropology': None,\n",
    " '5.5. geography': None,\n",
    " '5.6. recommender systems': ['recommendation engine', 'collaborative filtering'],\n",
    " '5.7. interpretation': ['explaining'],\n",
    " '5.8.1. ontology': ['knowledge graph', 'knowledge base'],\n",
    " '5.8.2. reasoning': ['expert system', 'inference engine', 'production system', 'if-then rules'],\n",
    " '5.8.3. robotics': ['autonomous device'],\n",
    " '5.8.4. belief network': ['bayesian network', 'decision network'],\n",
    " '5.8.5. fuzzy logics': ['fuzzy systems'],\n",
    " '5.9. bioinformatics': ['genomics, proteomics'],\n",
    " '5.10. sociology': ['social science']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "papers_file = 'papers.csv'\n",
    "papers = pd.read_csv(papers_file)\n",
    "abstracts = list(papers['abstract'])\n",
    "keywords = list(papers['keywords'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(topics_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "332"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(abstracts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clear_text(text, lowerize=True):\n",
    "    # чистим все, что не входит в задданый компайл\n",
    "    pat = re.compile(r'[^A-Za-z0-9 \\-\\n\\r.,;!?А-Яа-я]+')\n",
    "    cleared_text = re.sub(pat, ' ', text)\n",
    "\n",
    "    # переводим в нижний регистр\n",
    "    if lowerize:\n",
    "        cleared_text = cleared_text.lower()\n",
    "    # токенизируем текст\n",
    "    tokens = cleared_text.split()\n",
    "    return tokens\n",
    "\n",
    "def make_substrings(tokens, k=4):\n",
    "    # передаем подстроку из k токенов\n",
    "    for i in range(max(len(tokens) - k + 1, 1)):\n",
    "        yield ' '.join(tokens[i:i + k])\n",
    "\n",
    "def text_to_topic(texts, topics):\n",
    "    matrix = np.empty((0, len(topics)), float)\n",
    "    texts_tokens = [clear_text(t) for t in texts]\n",
    "    topics_syn = []\n",
    "    for topic_key in topics:\n",
    "        topic_syn = [' '.join(topic_key.split()[1:])]\n",
    "        if topics[topic_key] is not None:\n",
    "            topic_syn += topics[topic_key]\n",
    "        topics_syn.append(topic_syn)\n",
    "        \n",
    "    for text_tokens in texts_tokens:\n",
    "        ast = base.AST.get_ast(list(make_substrings(text_tokens)))\n",
    "        scores = []\n",
    "        for topic_syn in topics_syn:\n",
    "            score = 0\n",
    "            for syn in topic_syn:\n",
    "                score = max(ast.score(syn), score)\n",
    "            scores.append(score)\n",
    "        row = np.array(scores)    \n",
    "        matrix = np.append(matrix, [row], axis=0)\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.06173732, 0.12621883, 0.11083396, ..., 0.05980933, 0.11624258,\n",
       "        0.09911761],\n",
       "       [0.06713455, 0.1869032 , 0.15072022, ..., 0.03909487, 0.17370178,\n",
       "        0.15429046],\n",
       "       [0.06591225, 0.18058078, 0.11057313, ..., 0.0490205 , 0.28586136,\n",
       "        0.14559464],\n",
       "       ...,\n",
       "       [0.07199158, 0.08935749, 0.11486927, ..., 0.04479044, 0.14603816,\n",
       "        0.18010459],\n",
       "       [0.12977814, 0.21023098, 0.15786837, ..., 0.22196176, 0.33823234,\n",
       "        0.07230678],\n",
       "       [0.16877715, 0.10383607, 0.09990654, ..., 0.04067776, 0.2142774 ,\n",
       "        0.06238691]])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_text = text_to_topic(abstracts, topics_dict)\n",
    "m_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clear_keyword(keyword, lowerize=True):\n",
    "    # чистим все, что не входит в задданый компайл\n",
    "    pat = re.compile(r'[^A-Za-z0-9 \\-\\n\\r.,;\\'!?А-Яа-я]+')\n",
    "    cleared_keyword = re.sub(pat, ' ', keyword)\n",
    "\n",
    "    # переводим в нижний регистр\n",
    "    if lowerize:\n",
    "        cleared_keyword = cleared_keyword.lower()\n",
    "    return cleared_keyword.split(', ')\n",
    "\n",
    "def keyw_to_topic(keywords, topics):\n",
    "    matrix = np.empty((0, len(topics)), float)\n",
    "    keywords_tokens = []\n",
    "    for t in keywords:\n",
    "        if t is not np.nan:\n",
    "            keywords_tokens.append(clear_keyword(t))\n",
    "        else:\n",
    "            keywords_tokens.append([''])\n",
    "    topics_syn = []\n",
    "    for topic_key in topics:\n",
    "        topic_syn = [' '.join(topic_key.split()[1:])]\n",
    "        if topics[topic_key] is not None:\n",
    "            topic_syn += topics[topic_key]\n",
    "        topics_syn.append(topic_syn)\n",
    "\n",
    "    for keyword_tokens in keywords_tokens:\n",
    "        scores = []\n",
    "        for topic_syn in topics_syn:\n",
    "            score = 0\n",
    "            for syn in topic_syn:\n",
    "                for keyword in keyword_tokens:\n",
    "                    a = set(syn.split())\n",
    "                    b = set(keyword.split())\n",
    "                    c = len(a.intersection(b))/len(a.union(b))\n",
    "                    score = max(score, c)\n",
    "            scores.append(np.sqrt(score))\n",
    "        row = np.array(scores)    \n",
    "        matrix = np.append(matrix, [row], axis=0)\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.       , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [0.       , 0.4472136, 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       ...,\n",
       "       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ]])"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_keyw = keyw_to_topic(keywords, topics_dict)\n",
    "m_keyw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.04115821, 0.08414589, 0.07388931, ..., 0.03987288, 0.07749505,\n",
       "        0.06607841],\n",
       "       [0.04475637, 0.12460213, 0.10048015, ..., 0.02606325, 0.11580119,\n",
       "        0.10286031],\n",
       "       [0.0439415 , 0.26945838, 0.07371542, ..., 0.03268033, 0.19057424,\n",
       "        0.0970631 ],\n",
       "       ...,\n",
       "       [0.04799439, 0.05957166, 0.07657951, ..., 0.0298603 , 0.09735877,\n",
       "        0.12006973],\n",
       "       [0.08651876, 0.14015398, 0.10524558, ..., 0.14797451, 0.22548822,\n",
       "        0.04820452],\n",
       "       [0.1125181 , 0.06922405, 0.06660436, ..., 0.02711851, 0.1428516 ,\n",
       "        0.04159128]])"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m = 2/3 * m_text + 1/3 * m_keyw\n",
    "np.savetxt('m.txt', m)\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = np.loadtxt('m.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def topic_to_topic(R, threshold):\n",
    "    if threshold is not None:\n",
    "        ost = []\n",
    "        n_v = np.sum(R > threshold, axis=1)\n",
    "        for i in range(len(n_v)):\n",
    "            if n_v[i] == 0:\n",
    "                ost.append(i)\n",
    "        R = np.delete(R, ost, axis=0)\n",
    "        n_v = np.delete(n_v, ost)\n",
    "        w = 1 / n_v\n",
    "        A = R.T.dot(np.diag(w)).dot(R)\n",
    "        return A, ost\n",
    "    else:\n",
    "        A = R.T.dot(R)\n",
    "    return A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[277]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A, ost = topic_to_topic(m, threshold=0.3)\n",
    "ost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lapin import lapin\n",
    "from faddis import faddis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster contribution is too small\n"
     ]
    }
   ],
   "source": [
    "lapin_A = lapin(A)\n",
    "u_matrix, contrib = faddis(lapin_A)\n",
    "np.savetxt('u_matrix.txt', u_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.00742309, 0.        , 0.        , ..., 0.0027268 , 0.17279547,\n",
       "        0.01639823],\n",
       "       [0.        , 0.00240153, 0.        , ..., 0.04348909, 0.        ,\n",
       "        0.06986134],\n",
       "       [0.00344547, 0.        , 0.        , ..., 0.0565049 , 0.        ,\n",
       "        0.        ],\n",
       "       ...,\n",
       "       [0.        , 0.        , 0.02617854, ..., 0.18438576, 0.        ,\n",
       "        0.00594685],\n",
       "       [0.        , 0.        , 0.        , ..., 0.00217406, 0.        ,\n",
       "        0.        ],\n",
       "       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,\n",
       "        0.00372325]])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.loadtxt('u_matrix.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "codes = np.array([t.split()[0] for t in topics_dict])\n",
    "topices = np.array([' '.join(t.split()[1:]) for t in topics_dict])\n",
    "cluster_matrix = pd.DataFrame(u_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 3</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 4</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 5</th>\n",
       "      <th></th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster 8</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 9</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 10</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 11</th>\n",
       "      <th></th>\n",
       "      <th>Cluster 12</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.9.other clustering</td>\n",
       "      <td>0.935468</td>\n",
       "      <td>1.5.2.mixture</td>\n",
       "      <td>0.588406</td>\n",
       "      <td>2.7.temporal data</td>\n",
       "      <td>0.478999</td>\n",
       "      <td>4.5.4.bootstrap</td>\n",
       "      <td>0.815775</td>\n",
       "      <td>5.1.classification</td>\n",
       "      <td>0.722299</td>\n",
       "      <td>...</td>\n",
       "      <td>3.8.1.non-negative decomposition</td>\n",
       "      <td>0.643757</td>\n",
       "      <td>3.9.anomaly detection</td>\n",
       "      <td>0.618479</td>\n",
       "      <td>1.1.4.affinity propagation</td>\n",
       "      <td>0.550502</td>\n",
       "      <td>3.7.1.biplot</td>\n",
       "      <td>0.689529</td>\n",
       "      <td>3.4.2.principal component analysis</td>\n",
       "      <td>0.528809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.3.2.divisive</td>\n",
       "      <td>0.149594</td>\n",
       "      <td>4.1.1.mixture of distributions</td>\n",
       "      <td>0.459158</td>\n",
       "      <td>2.6.spatial data</td>\n",
       "      <td>0.440132</td>\n",
       "      <td>5.3.psychology</td>\n",
       "      <td>0.232690</td>\n",
       "      <td>1.3.3.decision tree</td>\n",
       "      <td>0.509320</td>\n",
       "      <td>...</td>\n",
       "      <td>2.1.6.single linkage</td>\n",
       "      <td>0.601147</td>\n",
       "      <td>2.10.3.edge detection</td>\n",
       "      <td>0.427353</td>\n",
       "      <td>2.1.1.additive clustering</td>\n",
       "      <td>0.518403</td>\n",
       "      <td>4.2.geometry</td>\n",
       "      <td>0.335729</td>\n",
       "      <td>3.8.2.archetype analysis</td>\n",
       "      <td>0.476016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.1.6.single linkage</td>\n",
       "      <td>0.122258</td>\n",
       "      <td>4.1.3.maximum likelihood</td>\n",
       "      <td>0.407998</td>\n",
       "      <td>2.11.5.data normalization</td>\n",
       "      <td>0.389796</td>\n",
       "      <td>2.2.distance</td>\n",
       "      <td>0.227840</td>\n",
       "      <td>3.3.1.discriminant analysis</td>\n",
       "      <td>0.237562</td>\n",
       "      <td>...</td>\n",
       "      <td>4.5.3.leave-one-out</td>\n",
       "      <td>0.224817</td>\n",
       "      <td>2.1.3.community detection</td>\n",
       "      <td>0.383785</td>\n",
       "      <td>5.5.geography</td>\n",
       "      <td>0.336215</td>\n",
       "      <td>4.5.4.bootstrap</td>\n",
       "      <td>0.242833</td>\n",
       "      <td>3.3.1.discriminant analysis</td>\n",
       "      <td>0.451014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.6.recommender systems</td>\n",
       "      <td>0.111728</td>\n",
       "      <td>1.5.4.markov models</td>\n",
       "      <td>0.287464</td>\n",
       "      <td>2.11.3.data coding</td>\n",
       "      <td>0.385837</td>\n",
       "      <td>5.4.antropology</td>\n",
       "      <td>0.176554</td>\n",
       "      <td>4.1.2.bayesian</td>\n",
       "      <td>0.206034</td>\n",
       "      <td>...</td>\n",
       "      <td>2.1.1.additive clustering</td>\n",
       "      <td>0.187462</td>\n",
       "      <td>1.5.1.density function</td>\n",
       "      <td>0.260523</td>\n",
       "      <td>5.4.antropology</td>\n",
       "      <td>0.214957</td>\n",
       "      <td>5.7.interpretation</td>\n",
       "      <td>0.213909</td>\n",
       "      <td>3.4.3.correspondence analysis</td>\n",
       "      <td>0.425522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.6.2.false positive</td>\n",
       "      <td>0.098004</td>\n",
       "      <td>1.5.3.expectation maximization</td>\n",
       "      <td>0.273233</td>\n",
       "      <td>4.5.5.data generation</td>\n",
       "      <td>0.370608</td>\n",
       "      <td>3.10.3.association index</td>\n",
       "      <td>0.161669</td>\n",
       "      <td>3.8.2.archetype analysis</td>\n",
       "      <td>0.164551</td>\n",
       "      <td>...</td>\n",
       "      <td>2.10.3.edge detection</td>\n",
       "      <td>0.136691</td>\n",
       "      <td>2.9.3.information retrieval</td>\n",
       "      <td>0.211468</td>\n",
       "      <td>1.1.5.block-modeling</td>\n",
       "      <td>0.195655</td>\n",
       "      <td>2.9.2.topic modeling</td>\n",
       "      <td>0.213239</td>\n",
       "      <td>3.4.4.independent component</td>\n",
       "      <td>0.150607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3.7.2.dendrogram</td>\n",
       "      <td>0.095371</td>\n",
       "      <td>1.5.1.density function</td>\n",
       "      <td>0.269427</td>\n",
       "      <td>1.1.6.semi-supervised clustering</td>\n",
       "      <td>0.318376</td>\n",
       "      <td>5.10.sociology</td>\n",
       "      <td>0.148793</td>\n",
       "      <td>3.4.2.principal component analysis</td>\n",
       "      <td>0.160634</td>\n",
       "      <td>...</td>\n",
       "      <td>3.1.ranking</td>\n",
       "      <td>0.135481</td>\n",
       "      <td>4.1.2.bayesian</td>\n",
       "      <td>0.195009</td>\n",
       "      <td>5.8.5.fuzzy logics</td>\n",
       "      <td>0.184386</td>\n",
       "      <td>2.2.distance</td>\n",
       "      <td>0.210723</td>\n",
       "      <td>1.5.2.mixture</td>\n",
       "      <td>0.136744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.4.fuzzy clustering</td>\n",
       "      <td>0.086878</td>\n",
       "      <td>2.1.5.maximum spanning tree</td>\n",
       "      <td>0.101011</td>\n",
       "      <td>3.3.5.regularizer</td>\n",
       "      <td>0.083309</td>\n",
       "      <td>5.8.1.ontology</td>\n",
       "      <td>0.143404</td>\n",
       "      <td>3.4.3.correspondence analysis</td>\n",
       "      <td>0.138354</td>\n",
       "      <td>...</td>\n",
       "      <td>1.1.3.self organised map</td>\n",
       "      <td>0.127046</td>\n",
       "      <td>2.9.4.text mining</td>\n",
       "      <td>0.154137</td>\n",
       "      <td>3.4.3.correspondence analysis</td>\n",
       "      <td>0.148514</td>\n",
       "      <td>3.3.2.support vector machine</td>\n",
       "      <td>0.191805</td>\n",
       "      <td>1.2.one cluster</td>\n",
       "      <td>0.105760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2.1.5.maximum spanning tree</td>\n",
       "      <td>0.082138</td>\n",
       "      <td>1.1.5.block-modeling</td>\n",
       "      <td>0.079934</td>\n",
       "      <td>2.1.8.kernel function</td>\n",
       "      <td>0.081287</td>\n",
       "      <td>2.10.2.skeleton</td>\n",
       "      <td>0.137882</td>\n",
       "      <td>5.8.4.belief network</td>\n",
       "      <td>0.102451</td>\n",
       "      <td>...</td>\n",
       "      <td>5.8.5.fuzzy logics</td>\n",
       "      <td>0.101660</td>\n",
       "      <td>4.1.3.maximum likelihood</td>\n",
       "      <td>0.145702</td>\n",
       "      <td>1.9.other clustering</td>\n",
       "      <td>0.147405</td>\n",
       "      <td>1.1.1.k-means</td>\n",
       "      <td>0.172795</td>\n",
       "      <td>3.6.biclustering</td>\n",
       "      <td>0.098889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4.1.1.mixture of distributions</td>\n",
       "      <td>0.074715</td>\n",
       "      <td>4.3.optimisation</td>\n",
       "      <td>0.070607</td>\n",
       "      <td>3.9.anomaly detection</td>\n",
       "      <td>0.061323</td>\n",
       "      <td>3.12.formal concept</td>\n",
       "      <td>0.135228</td>\n",
       "      <td>1.5.2.mixture</td>\n",
       "      <td>0.083569</td>\n",
       "      <td>...</td>\n",
       "      <td>3.4.2.principal component analysis</td>\n",
       "      <td>0.100268</td>\n",
       "      <td>5.9.bioinformatics</td>\n",
       "      <td>0.128393</td>\n",
       "      <td>3.10.2.comparing partitions</td>\n",
       "      <td>0.144406</td>\n",
       "      <td>2.8.three-way</td>\n",
       "      <td>0.142754</td>\n",
       "      <td>4.4.cluster structures</td>\n",
       "      <td>0.081083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.12.formal concept</td>\n",
       "      <td>0.073771</td>\n",
       "      <td>2.9.2.topic modeling</td>\n",
       "      <td>0.064499</td>\n",
       "      <td>2.5.incomplete</td>\n",
       "      <td>0.057663</td>\n",
       "      <td>4.3.optimisation</td>\n",
       "      <td>0.123337</td>\n",
       "      <td>4.1.1.mixture of distributions</td>\n",
       "      <td>0.083138</td>\n",
       "      <td>...</td>\n",
       "      <td>1.1.2.number of clusters</td>\n",
       "      <td>0.072641</td>\n",
       "      <td>4.1.4.markov-chain monte carlo methods</td>\n",
       "      <td>0.101831</td>\n",
       "      <td>3.3.3.nearest neighbor</td>\n",
       "      <td>0.131248</td>\n",
       "      <td>1.5.4.markov models</td>\n",
       "      <td>0.133046</td>\n",
       "      <td>2.8.three-way</td>\n",
       "      <td>0.074301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.008342</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.007951</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.007582</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.007426</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.007208</td>\n",
       "      <td>...</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.006081</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.005889</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.005237</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.005040</td>\n",
       "      <td>Contribution</td>\n",
       "      <td>0.004558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Cluster 1                                 Cluster 2  \\\n",
       "0             1.9.other clustering  0.935468                   1.5.2.mixture   \n",
       "1                   1.3.2.divisive  0.149594  4.1.1.mixture of distributions   \n",
       "2             2.1.6.single linkage  0.122258        4.1.3.maximum likelihood   \n",
       "3          5.6.recommender systems  0.111728             1.5.4.markov models   \n",
       "4             4.6.2.false positive  0.098004  1.5.3.expectation maximization   \n",
       "5                 3.7.2.dendrogram  0.095371          1.5.1.density function   \n",
       "6             1.4.fuzzy clustering  0.086878     2.1.5.maximum spanning tree   \n",
       "7      2.1.5.maximum spanning tree  0.082138            1.1.5.block-modeling   \n",
       "8   4.1.1.mixture of distributions  0.074715                4.3.optimisation   \n",
       "9              3.12.formal concept  0.073771            2.9.2.topic modeling   \n",
       "10                    Contribution  0.008342                    Contribution   \n",
       "\n",
       "                                     Cluster 3            \\\n",
       "0   0.588406                 2.7.temporal data  0.478999   \n",
       "1   0.459158                  2.6.spatial data  0.440132   \n",
       "2   0.407998         2.11.5.data normalization  0.389796   \n",
       "3   0.287464                2.11.3.data coding  0.385837   \n",
       "4   0.273233             4.5.5.data generation  0.370608   \n",
       "5   0.269427  1.1.6.semi-supervised clustering  0.318376   \n",
       "6   0.101011                 3.3.5.regularizer  0.083309   \n",
       "7   0.079934             2.1.8.kernel function  0.081287   \n",
       "8   0.070607             3.9.anomaly detection  0.061323   \n",
       "9   0.064499                    2.5.incomplete  0.057663   \n",
       "10  0.007951                      Contribution  0.007582   \n",
       "\n",
       "                   Cluster 4                                     Cluster 5  \\\n",
       "0            4.5.4.bootstrap  0.815775                  5.1.classification   \n",
       "1             5.3.psychology  0.232690                 1.3.3.decision tree   \n",
       "2               2.2.distance  0.227840         3.3.1.discriminant analysis   \n",
       "3            5.4.antropology  0.176554                      4.1.2.bayesian   \n",
       "4   3.10.3.association index  0.161669            3.8.2.archetype analysis   \n",
       "5             5.10.sociology  0.148793  3.4.2.principal component analysis   \n",
       "6             5.8.1.ontology  0.143404       3.4.3.correspondence analysis   \n",
       "7            2.10.2.skeleton  0.137882                5.8.4.belief network   \n",
       "8        3.12.formal concept  0.135228                       1.5.2.mixture   \n",
       "9           4.3.optimisation  0.123337      4.1.1.mixture of distributions   \n",
       "10              Contribution  0.007426                        Contribution   \n",
       "\n",
       "              ...                           Cluster 8            \\\n",
       "0   0.722299  ...    3.8.1.non-negative decomposition  0.643757   \n",
       "1   0.509320  ...                2.1.6.single linkage  0.601147   \n",
       "2   0.237562  ...                 4.5.3.leave-one-out  0.224817   \n",
       "3   0.206034  ...           2.1.1.additive clustering  0.187462   \n",
       "4   0.164551  ...               2.10.3.edge detection  0.136691   \n",
       "5   0.160634  ...                         3.1.ranking  0.135481   \n",
       "6   0.138354  ...            1.1.3.self organised map  0.127046   \n",
       "7   0.102451  ...                  5.8.5.fuzzy logics  0.101660   \n",
       "8   0.083569  ...  3.4.2.principal component analysis  0.100268   \n",
       "9   0.083138  ...            1.1.2.number of clusters  0.072641   \n",
       "10  0.007208  ...                        Contribution  0.006081   \n",
       "\n",
       "                                 Cluster 9            \\\n",
       "0                    3.9.anomaly detection  0.618479   \n",
       "1                    2.10.3.edge detection  0.427353   \n",
       "2                2.1.3.community detection  0.383785   \n",
       "3                   1.5.1.density function  0.260523   \n",
       "4              2.9.3.information retrieval  0.211468   \n",
       "5                           4.1.2.bayesian  0.195009   \n",
       "6                        2.9.4.text mining  0.154137   \n",
       "7                 4.1.3.maximum likelihood  0.145702   \n",
       "8                       5.9.bioinformatics  0.128393   \n",
       "9   4.1.4.markov-chain monte carlo methods  0.101831   \n",
       "10                            Contribution  0.005889   \n",
       "\n",
       "                       Cluster 10                               Cluster 11  \\\n",
       "0      1.1.4.affinity propagation   0.550502                  3.7.1.biplot   \n",
       "1       2.1.1.additive clustering   0.518403                  4.2.geometry   \n",
       "2                   5.5.geography   0.336215               4.5.4.bootstrap   \n",
       "3                 5.4.antropology   0.214957            5.7.interpretation   \n",
       "4            1.1.5.block-modeling   0.195655          2.9.2.topic modeling   \n",
       "5              5.8.5.fuzzy logics   0.184386                  2.2.distance   \n",
       "6   3.4.3.correspondence analysis   0.148514  3.3.2.support vector machine   \n",
       "7            1.9.other clustering   0.147405                 1.1.1.k-means   \n",
       "8     3.10.2.comparing partitions   0.144406                 2.8.three-way   \n",
       "9          3.3.3.nearest neighbor   0.131248           1.5.4.markov models   \n",
       "10                   Contribution   0.005237                  Contribution   \n",
       "\n",
       "                                        Cluster 12               \n",
       "0     0.689529  3.4.2.principal component analysis     0.528809  \n",
       "1     0.335729            3.8.2.archetype analysis     0.476016  \n",
       "2     0.242833         3.3.1.discriminant analysis     0.451014  \n",
       "3     0.213909       3.4.3.correspondence analysis     0.425522  \n",
       "4     0.213239         3.4.4.independent component     0.150607  \n",
       "5     0.210723                       1.5.2.mixture     0.136744  \n",
       "6     0.191805                     1.2.one cluster     0.105760  \n",
       "7     0.172795                    3.6.biclustering     0.098889  \n",
       "8     0.142754              4.4.cluster structures     0.081083  \n",
       "9     0.133046                       2.8.three-way     0.074301  \n",
       "10    0.005040                        Contribution     0.004558  \n",
       "\n",
       "[11 rows x 24 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dic = {}\n",
    "for num_cl in cluster_matrix:\n",
    "    val = cluster_matrix[num_cl].sort_values(ascending=False)[:10]\n",
    "    ind = val.index\n",
    "    cod = codes[ind]\n",
    "    top = topices[ind]\n",
    "    s = []\n",
    "    for i in range(len(cod)):\n",
    "        s.append(cod[i] + top[i])\n",
    "    s.append('Contribution')\n",
    "    dic['Cluster ' + str(num_cl + 1)] = s\n",
    "    dic[' ' * num_cl] = np.append(val.values, contrib[num_cl])\n",
    "d = pd.DataFrame(dic)\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.to_excel('clusters.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters = [(d.iloc[:1, [0, 1]], 'Other clustering'), (d.iloc[:3, [2, 3]], 'Probabilistic'),\n",
    "            (d.iloc[:6, [4, 5]], 'Data structure'), (d.iloc[:9, [6, 7]], 'Application'),\n",
    "            (d.iloc[:4, [8, 9]], 'Tree'), (d.iloc[:4, [10, 11]], 'Cluster-analysis methods'),\n",
    "            (d.iloc[:4, [12, 13]], 'Regression'), (d.iloc[:3, [16, 17]], 'Detection'),\n",
    "            (d.iloc[:5, [22, 23]], 'Hidden factors +DA')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topics = []\n",
    "for topic_key in topics_dict:\n",
    "    topics.append(' '.join(topic_key.split()[1:]))\n",
    "len(topics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Other clustering</th>\n",
       "      <th>Probabilistic</th>\n",
       "      <th>Data structure</th>\n",
       "      <th>Application</th>\n",
       "      <th>Tree</th>\n",
       "      <th>Cluster-analysis methods</th>\n",
       "      <th>Regression</th>\n",
       "      <th>Detection</th>\n",
       "      <th>Hidden factors +DA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transfer distance between partitions</td>\n",
       "      <td>Latent class CUB models</td>\n",
       "      <td>Generalized mixture models, semi-supervised le...</td>\n",
       "      <td>Power analysis for the bootstrap likelihood ra...</td>\n",
       "      <td>Supervised learning via smoothed Polya trees</td>\n",
       "      <td>Orthogonal nonnegative matrix tri-factorizatio...</td>\n",
       "      <td>Maximum likelihood estimation of Gaussian mixt...</td>\n",
       "      <td>Infinite Dirichlet mixture models learning via...</td>\n",
       "      <td>Orthogonal rotation in PCAMIX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A review of robust clustering methods</td>\n",
       "      <td>On mixtures of skew normal and skew \\(t\\)-dist...</td>\n",
       "      <td>Equi-Clustream: a framework for clustering tim...</td>\n",
       "      <td>Rethinking an ROC partial area index for evalu...</td>\n",
       "      <td>Robust kernel principal component analysis and...</td>\n",
       "      <td>Mutual information, phi-squared and model-base...</td>\n",
       "      <td>A fuzzy approach to robust regression clustering</td>\n",
       "      <td>Plastic card fraud detection using peer group ...</td>\n",
       "      <td>Robust kernel principal component analysis and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mixture model averaging for clustering</td>\n",
       "      <td>Latent class model with conditional dependency...</td>\n",
       "      <td>A fragmented-periodogram approach for clusteri...</td>\n",
       "      <td>Inference for robust canonical variate analysis</td>\n",
       "      <td>Predicting partial customer churn using Markov...</td>\n",
       "      <td>Mixture model averaging for clustering</td>\n",
       "      <td>Model-based regression clustering for high-dim...</td>\n",
       "      <td>Robust clustering around regression lines with...</td>\n",
       "      <td>Multiple taxicab correspondence analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Functional data clustering: a survey</td>\n",
       "      <td>A constrained robust proposal for mixture mode...</td>\n",
       "      <td>Directional co-clustering</td>\n",
       "      <td>Multiple imputation in principal component ana...</td>\n",
       "      <td>Threshold optimization for classification in i...</td>\n",
       "      <td>Clustering of imbalanced high-dimensional medi...</td>\n",
       "      <td>Exploration of the variability of variable sel...</td>\n",
       "      <td>Outlier detection and robust covariance estima...</td>\n",
       "      <td>Multiple imputation in principal component ana...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Robust clustering for functional data based on...</td>\n",
       "      <td>Eigenvalues and constraints in mixture modelin...</td>\n",
       "      <td>Model-based clustering of time series in group...</td>\n",
       "      <td>Exploration of the variability of variable sel...</td>\n",
       "      <td>Marginal and simultaneous predictive classific...</td>\n",
       "      <td>An update algorithm for restricted random walk...</td>\n",
       "      <td>Regularized fuzzy clusterwise ridge regression</td>\n",
       "      <td>Detection of multivariate outliers in business...</td>\n",
       "      <td>Constrained principal component analysis of st...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Data clustering based on principal curves</td>\n",
       "      <td>Finite mixture biclustering of discrete type m...</td>\n",
       "      <td>Dense traffic flow patterns mining in bi-direc...</td>\n",
       "      <td>A Monte Carlo evaluation of three methods to d...</td>\n",
       "      <td>Classification using sequential order statistics</td>\n",
       "      <td>Probabilistic assessment of model-based cluste...</td>\n",
       "      <td>Least squares estimation of linear regression ...</td>\n",
       "      <td>The value of the last digit: statistical fraud...</td>\n",
       "      <td>Principal differential analysis of the Aneuris...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A mixture of generalized hyperbolic factor ana...</td>\n",
       "      <td>Finite mixture-of-gamma distributions: estimat...</td>\n",
       "      <td>Unifying data units and models in (co-)clustering</td>\n",
       "      <td>Classification in music research</td>\n",
       "      <td>Dichotomic lattices and local discretization f...</td>\n",
       "      <td>Infinite Dirichlet mixture models learning via...</td>\n",
       "      <td>Lasso-constrained regression analysis for inte...</td>\n",
       "      <td>A robust approach to model-based classificatio...</td>\n",
       "      <td>Principal component analysis for probabilistic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tag SNP selection based on clustering accordin...</td>\n",
       "      <td>Finite mixture of regression models for censor...</td>\n",
       "      <td>Simulating mixtures of multivariate data with ...</td>\n",
       "      <td>Two local dissimilarity measures for weighted ...</td>\n",
       "      <td>A new approach for determining the prior proba...</td>\n",
       "      <td>Tag SNP selection based on clustering accordin...</td>\n",
       "      <td>Linear regression for numeric symbolic variabl...</td>\n",
       "      <td>A robust approach to model-based classificatio...</td>\n",
       "      <td>Discriminant analysis for discrete variables d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A tail dependence-based dissimilarity measure ...</td>\n",
       "      <td>From here to infinity: sparse finite versus Di...</td>\n",
       "      <td>Latent drop-out based transitions in linear qu...</td>\n",
       "      <td>Ensemble of optimal trees, random forest and r...</td>\n",
       "      <td>T3C: improving a decision tree classification ...</td>\n",
       "      <td>Factor probabilistic distance clustering (FPDC...</td>\n",
       "      <td>Model-based clustering and segmentation of tim...</td>\n",
       "      <td>Robust estimation of efficient mean–variance f...</td>\n",
       "      <td>Correspondence analysis of textual data involv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Estimating common principal components in high...</td>\n",
       "      <td>Probabilistic clustering via Pareto solutions ...</td>\n",
       "      <td>Asymptotic comparison of semi-supervised and s...</td>\n",
       "      <td>Tree-structured modelling of categorical predi...</td>\n",
       "      <td>Bayesian unsupervised classification framework...</td>\n",
       "      <td>Functional data clustering: a survey</td>\n",
       "      <td>Tree-structured modelling of categorical predi...</td>\n",
       "      <td>Evaluation of the evolution of relationships b...</td>\n",
       "      <td>Parametric classification with soft labels usi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Other clustering  \\\n",
       "0               Transfer distance between partitions   \n",
       "1              A review of robust clustering methods   \n",
       "2             Mixture model averaging for clustering   \n",
       "3               Functional data clustering: a survey   \n",
       "4  Robust clustering for functional data based on...   \n",
       "5          Data clustering based on principal curves   \n",
       "6  A mixture of generalized hyperbolic factor ana...   \n",
       "7  Tag SNP selection based on clustering accordin...   \n",
       "8  A tail dependence-based dissimilarity measure ...   \n",
       "9  Estimating common principal components in high...   \n",
       "\n",
       "                                       Probabilistic  \\\n",
       "0                            Latent class CUB models   \n",
       "1  On mixtures of skew normal and skew \\(t\\)-dist...   \n",
       "2  Latent class model with conditional dependency...   \n",
       "3  A constrained robust proposal for mixture mode...   \n",
       "4  Eigenvalues and constraints in mixture modelin...   \n",
       "5  Finite mixture biclustering of discrete type m...   \n",
       "6  Finite mixture-of-gamma distributions: estimat...   \n",
       "7  Finite mixture of regression models for censor...   \n",
       "8  From here to infinity: sparse finite versus Di...   \n",
       "9  Probabilistic clustering via Pareto solutions ...   \n",
       "\n",
       "                                      Data structure  \\\n",
       "0  Generalized mixture models, semi-supervised le...   \n",
       "1  Equi-Clustream: a framework for clustering tim...   \n",
       "2  A fragmented-periodogram approach for clusteri...   \n",
       "3                          Directional co-clustering   \n",
       "4  Model-based clustering of time series in group...   \n",
       "5  Dense traffic flow patterns mining in bi-direc...   \n",
       "6  Unifying data units and models in (co-)clustering   \n",
       "7  Simulating mixtures of multivariate data with ...   \n",
       "8  Latent drop-out based transitions in linear qu...   \n",
       "9  Asymptotic comparison of semi-supervised and s...   \n",
       "\n",
       "                                         Application  \\\n",
       "0  Power analysis for the bootstrap likelihood ra...   \n",
       "1  Rethinking an ROC partial area index for evalu...   \n",
       "2    Inference for robust canonical variate analysis   \n",
       "3  Multiple imputation in principal component ana...   \n",
       "4  Exploration of the variability of variable sel...   \n",
       "5  A Monte Carlo evaluation of three methods to d...   \n",
       "6                   Classification in music research   \n",
       "7  Two local dissimilarity measures for weighted ...   \n",
       "8  Ensemble of optimal trees, random forest and r...   \n",
       "9  Tree-structured modelling of categorical predi...   \n",
       "\n",
       "                                                Tree  \\\n",
       "0       Supervised learning via smoothed Polya trees   \n",
       "1  Robust kernel principal component analysis and...   \n",
       "2  Predicting partial customer churn using Markov...   \n",
       "3  Threshold optimization for classification in i...   \n",
       "4  Marginal and simultaneous predictive classific...   \n",
       "5   Classification using sequential order statistics   \n",
       "6  Dichotomic lattices and local discretization f...   \n",
       "7  A new approach for determining the prior proba...   \n",
       "8  T3C: improving a decision tree classification ...   \n",
       "9  Bayesian unsupervised classification framework...   \n",
       "\n",
       "                            Cluster-analysis methods  \\\n",
       "0  Orthogonal nonnegative matrix tri-factorizatio...   \n",
       "1  Mutual information, phi-squared and model-base...   \n",
       "2             Mixture model averaging for clustering   \n",
       "3  Clustering of imbalanced high-dimensional medi...   \n",
       "4  An update algorithm for restricted random walk...   \n",
       "5  Probabilistic assessment of model-based cluste...   \n",
       "6  Infinite Dirichlet mixture models learning via...   \n",
       "7  Tag SNP selection based on clustering accordin...   \n",
       "8  Factor probabilistic distance clustering (FPDC...   \n",
       "9               Functional data clustering: a survey   \n",
       "\n",
       "                                          Regression  \\\n",
       "0  Maximum likelihood estimation of Gaussian mixt...   \n",
       "1   A fuzzy approach to robust regression clustering   \n",
       "2  Model-based regression clustering for high-dim...   \n",
       "3  Exploration of the variability of variable sel...   \n",
       "4     Regularized fuzzy clusterwise ridge regression   \n",
       "5  Least squares estimation of linear regression ...   \n",
       "6  Lasso-constrained regression analysis for inte...   \n",
       "7  Linear regression for numeric symbolic variabl...   \n",
       "8  Model-based clustering and segmentation of tim...   \n",
       "9  Tree-structured modelling of categorical predi...   \n",
       "\n",
       "                                           Detection  \\\n",
       "0  Infinite Dirichlet mixture models learning via...   \n",
       "1  Plastic card fraud detection using peer group ...   \n",
       "2  Robust clustering around regression lines with...   \n",
       "3  Outlier detection and robust covariance estima...   \n",
       "4  Detection of multivariate outliers in business...   \n",
       "5  The value of the last digit: statistical fraud...   \n",
       "6  A robust approach to model-based classificatio...   \n",
       "7  A robust approach to model-based classificatio...   \n",
       "8  Robust estimation of efficient mean–variance f...   \n",
       "9  Evaluation of the evolution of relationships b...   \n",
       "\n",
       "                                  Hidden factors +DA  \n",
       "0                      Orthogonal rotation in PCAMIX  \n",
       "1  Robust kernel principal component analysis and...  \n",
       "2           Multiple taxicab correspondence analysis  \n",
       "3  Multiple imputation in principal component ana...  \n",
       "4  Constrained principal component analysis of st...  \n",
       "5  Principal differential analysis of the Aneuris...  \n",
       "6  Principal component analysis for probabilistic...  \n",
       "7  Discriminant analysis for discrete variables d...  \n",
       "8  Correspondence analysis of textual data involv...  \n",
       "9  Parametric classification with soft labels usi...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_cl = {}\n",
    "for num, cluster_area in enumerate(clusters):\n",
    "    cluster = cluster_area[0]\n",
    "    area = cluster_area[1]\n",
    "    topics_ind = list(map(lambda x: topics.index(x.split('.')[-1]), cluster.iloc[:, 0]))\n",
    "    topics_rel = np.zeros(len(topics))\n",
    "    topics_rel[topics_ind] = cluster.iloc[:, 1].values\n",
    "    a = np.dot(m, topics_rel)/len(topics_ind)\n",
    "    top_abstracts = np.argsort(a)[:-11:-1]\n",
    "    top_rel = np.sort(a)[:-11:-1]\n",
    "    dict_cl[area] = papers['paper name'][top_abstracts].values\n",
    "    #dict_cl[area + ' relevance'] = top_rel\n",
    "dd = pd.DataFrame(dict_cl)\n",
    "dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "dd.to_excel('top_papers.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
